packages/oauth/src/index.ts

```diff
@@ -129,6 +129,87 @@ export class InstallProvider {
         authResult.botToken = queryResult.bot.token;
         authResult.botId = queryResult.bot.id;
         authResult.botUserId = queryResult.bot.userId;
+
+        // Token Rotation Enabled (Bot Token)
+        if (queryResult.bot.refreshToken !== undefined) {
+          authResult.botRefreshToken = queryResult.bot.refreshToken;
+          authResult.botTokenExpiresAt = queryResult.bot.expiresAt; // utc, seconds
+        }
+      }
+
+      // Token Rotation Enabled (User Token)
+      if (queryResult.user.refreshToken !== undefined) {
+        authResult.userRefreshToken = queryResult.user.refreshToken;
+        authResult.userTokenExpiresAt = queryResult.user.expiresAt; // utc, seconds
+      }
+
+      /*
+      * Token Rotation (Expiry Check + Refresh)
+      * The presence of `(bot|user)TokenExpiresAt` indicates having opted into token rotation.
+      * If the token has expired, or will expire within 2 hours, the token is refreshed and
+      * the `authResult` and `Installation` are updated with the new values.
+      */
+      if (authResult.botRefreshToken !== undefined || authResult.userRefreshToken !== undefined) {
+        const currentUTCSec = Math.floor(Date.now() / 1000); // seconds
+        const tokensToRefresh: string[] = [];
+        let client: WebClient;
+
+        if (authResult.botRefreshToken !== undefined && authResult.botTokenExpiresAt !== undefined) {
+          const botTokenExpiresIn = authResult.botTokenExpiresAt - currentUTCSec;
+          if (botTokenExpiresIn <= 7200) { // 2 hours
+            tokensToRefresh.push(authResult.botRefreshToken);
+          }
+        }
+
+        if (authResult.userRefreshToken !== undefined && authResult.userTokenExpiresAt !== undefined) {
+          const userTokenExpiresIn = authResult.userTokenExpiresAt - currentUTCSec;
+          if (userTokenExpiresIn <= 7200) { // 2 hours
+            tokensToRefresh.push(authResult.userRefreshToken);
+          }
+        }
+
+        if (tokensToRefresh.length > 0) {
+          const installationUpdates: any = { ...queryResult }; // TODO :: TS
+          client = new WebClient(undefined, this.clientOptions);
+
+          // TODO :: Concurrent Calls?
```

Comment:
Concurrent calls are fine as long as no call won't be skipped. Specifically, we cannot have delayed calls here as the calls can be forcibly terminated when running on a FaaS runtime (AWS Lambda, Google Cloud Functions, Azure Functions). We may want to have a flag option like processBeforeResponse in Bolt.

Reply:
Great point. Since it's only ever two calls, do you think implementing a concurrent strategy is necessary? Also, thoughts on how to move forward with the retry logic if a call fails?

Reply:
>do you think implementing a concurrent strategy is necessary?

Yes, I think that it's valuable for end-user experience and less 3 second timeouts. To avoid having an extra overhead, we would like to asynchronously do these refreshing API calls as it's not necessary for **this** incoming request handling. If the `oauth.v2.access` here takes 300-500 milliseconds in average, calling twice may take 1 second in a worst case scenario. This may not be a small overhead.

I haven't checked if this works but one possible idea would be having a private function like `async function refreshExpiringToken(tokensToRefresh)` and calling the method asynchronously (no `await`) by default. In addition to that, we can consider FaaS applications by calling the same method with `await` if `processBeforeResponse` is true.

>Also, thoughts on how to move forward with the retry logic if a call fails?

It depends on the error. If the error is recoverable (e.g., connection failure, timeouts), we can retry with the same set of parameters. If the response has an error code like invalid parameters or invalid auth, of course, we don't need to retry the call anymore.